### PR TITLE
[DOCS-315] - State docs updates to support updated CQL model

### PR DESCRIPTION
**Pull request by jimmyjames**
This PR better documents details of State and Atomic State so that developers have a better understanding of what is happening, and can make more informed decisions about using using state.

It also reorganizes the content a bit and adds information that was missing, such as the return object type when getting values out of state. 

@rappleg please review or delegate as appropriate, thanks!

### CODE REVIEW

File: smartapp-developers-guide/state.rst [hunk 1]
```diff
@@ -3,7 +3,14 @@
...
 ====================
 
-SmartApps and Device Handlers are all provided a ``state`` variable that will allow you to store data across executions.
+SmartApps and Device Handlers execute in response to various events or schedules; they are not continuously running.
+Since each execution is executed independently, it has no information regarding previous executions.
+This is often adequate for most SmartApps or Device Handlers, but sometimes, they need to remember information across executions.
+
+For this reason, SmartApps and Device Handlers can store small amounts of data using a map-like API, and retrieve this data in later executions.
+
+SmartApps can persist and retrieve data one of two ways - using the built-in ``state`` or ``atomicState`` objects (the details and difference between these implementations is discussed in detail in this document).
```
Comment (unixbeast) [inline smartapp-developers-guide/state.rst]: should be "are discussed in detail"

**Comment by unixbeast** [comment]
👍  looks good to me. Should be good to merge after @rappleg gives his ok.

**Comment by jimmyjames** [comment]
@rappleg or @Bijnagte can one of you review this for technical accuracy, or delegate to someone who can? We have some other changes that are dependent on getting this.

File: smartapp-developers-guide/state.rst [hunk 2]
```diff
@@ -14,203 +21,456 @@ SmartApps and Device Handlers are all provided a ``state`` variable that will al
...
+The reasons for this are:
+
+#. State is easier to work with, since it supports ``java.util.Map``.
+#. State is more performant than Atomic State, since it does not read or write to external storage during SmartApp execution.
+
+You may need to use Atomic State if code that updates a value in State may execute at the same time as another instance of the same SmartApp, updating the same State key, as discussed :ref:`here <state_potential_race_conditions>`.
+
+.. important::
+
+    The most important thing to note is to **never use both Atomic State and State in the same SmartApp. This can't be emphasized enough - doing so may result in data inconsistency, data corruption, or even data loss.**
```
Comment (rappleg) [inline smartapp-developers-guide/state.rst]: Glad to see this being spelled out again explicitly.

**Comment by rappleg** [comment]
A lot of good info here and looks good to me.  Sorry for taking so long to get back to this one.

👍

**Pull request merged**